### PR TITLE
Balance resolution ordering

### PR DESCRIPTION
**Pull request by enjeyw**
**Describe the Pull Request**

Changing the order in which balance checks are made, and moving it inside the create credit transfer object, as well as a bunch of other changes to the vendor reimbursement api

**Todo**

- [ ] Link relevant [trello card](https://trello.com/b/PA2LhlOh/sempo-dev) or [related issue](https://github.com/teamsempo/SempoBlockchain/issues) to the pull request
- [ ] Request review from relevant @person or team
- [ ] QA your pull request. This includes running the app, login, testing changes etc.
- [ ] Bump [backend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/config.py#L7) and/or [frontend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/app/package.json#L3) versions following Semver

### CODE REVIEW

File: app/server/api/credit_transfer_api.py [hunk 1]
```diff
@@ -145,11 +145,21 @@ def put(self, credit_transfer_id):
             }
             return make_response(jsonify(response_object)), 400
 
-        if action == 'COMPLETE':
-            credit_transfer.resolve_as_complete_and_trigger_blockchain()
+        try:
```
Comment (enjeyw) [inline app/server/api/credit_transfer_api.py:148]: literally just execption coverage to stop it from 500-ing

File: app/server/api/vendor_payout_api.py [hunk 2]
```diff
@@ -69,35 +69,39 @@ def post(self):
...
-                datetime.today().strftime('%Y-%m-%d'),
-                (datetime.today() + timedelta(days=7)).strftime('%Y-%m-%d'),
-                transfer.id,
-                cents_to_dollars(v.balance),
-                '',
-                '',
-            ])
+            withdrawal_amount = Decimal(v._balance_wei or 0) / Decimal(1e16)
+
+            if withdrawal_amount > 0:
```
Comment (enjeyw) [inline app/server/api/vendor_payout_api.py:95]: only make a withdrawal if theres funds to withdraw

File: app/server/api/vendor_payout_api.py [hunk 3]
```diff
@@ -26,8 +26,10 @@ def post(self):
         # Process post data
         post_data = request.get_json()
         account_ids = []
+        relist_existing = True
         if post_data:
             account_ids = post_data.get('accounts', [])
+            relist_existing = post_data.get('relist_existing', True)
```
Comment (enjeyw) [inline app/server/api/vendor_payout_api.py:32]: optionally return previous withdrawal data (without creating new withdrawals)

File: app/server/api/vendor_payout_api.py [hunk 4]
```diff
@@ -65,58 +67,74 @@ def post(self):
...
-                send_user=v.primary_user,
-                sender_transfer_account=v,
-                transfer_mode=TransferModeEnum.INTERNAL,
-                require_sender_approved=False,
-                automatically_resolve_complete=False,
-            )
-
-            db.session.flush()
+            if relist_existing:
+                withdrawals = (CreditTransfer.query
```
Comment (enjeyw) [inline app/server/api/vendor_payout_api.py:88]: add the previous withdrawals to the list

File: app/server/api/vendor_payout_api.py [hunk 5]
```diff
@@ -136,62 +154,69 @@ def post(self):
...
             reader = csv.DictReader(f)
+
         transfers = []
         for line in reader:
-            vendor = db.session.query(TransferAccount).filter(TransferAccount.id == line['ID']).first()
-            if not vendor.primary_user.has_vendor_role:
-                response_object = {
-                    'message': str(f'{vendor} is not a vendor!'),
-                }
-                return make_response(jsonify(response_object)), 400
```
Comment (enjeyw) [inline app/server/api/vendor_payout_api.py:146]: the vendor wasn't used anywhere and the 'ID' column was confusing in the spreadsheet (is it a vendor id or a transfer id), so removed

File: app/server/api/vendor_payout_api.py [hunk 6]
```diff
@@ -136,62 +154,69 @@ def post(self):
...
-                response_object = {
-                    'message': f'Tranfer with ID {line["Transaction ID"]} not found!'
-                }
-                return make_response(jsonify(response_object)), 400
-
-            if round(transfer.transfer_amount) != round(dollars_to_cents(line["UnitAmount"])):
-                response_object = {
-                    'message': f'Invalid transfer amount!'
-                }
-                return make_response(jsonify(response_object)), 400
```
Comment (enjeyw) [inline app/server/api/vendor_payout_api.py:160]: the problem with returning a 400 (or 500ing as per earlier), is that it creates a missleading response - if this transfer is the 4th one in the list, then the previous 3 will have already succeeded

File: app/server/api/vendor_payout_api.py [hunk 7]
```diff
@@ -136,62 +154,69 @@ def post(self):
...
+
+            try:
+                if line['Payment Has Been Made'].upper() == 'TRUE' and line['Bank Payment Date']:
                     transfer.resolve_as_complete_and_trigger_blockchain()
+                    message = 'Transfer Success'
+                elif line['Payment Has Been Made'] == 'FALSE':
+                    transfer.resolve_as_rejected()
+                    message = 'Transfer Rejected'
+            except Exception as e:
+                message = str(e)
```
Comment (enjeyw) [inline app/server/api/vendor_payout_api.py:189]: catching failures for reporting

File: app/server/models/credit_transfer.py [hunk 8]
```diff
@@ -237,12 +244,19 @@ def resolve_as_complete_and_trigger_blockchain(
...
-            raise Exception(f'Transfer resolve function called multiple times for transaction {self.id}')
-        self.check_sender_transfer_limits()
+            raise Exception(f'Resolve called multiple times for transfer {self.id}')
+        try:
+            self.check_sender_transfer_limits()
+        except TransferLimitError as e:
+            # Sempo admins can always bypass limits, allowing for things like emergency moving of funds etc
+            if hasattr(g, 'user') and AccessControl.has_suffient_role(g.user.roles, {'ADMIN': 'sempoadmin'}):
+                self.add_message(f'Warning: {e}')
+            else:
```
Comment (enjeyw) [inline app/server/models/credit_transfer.py:254]: allow sempoadmins to force a transfer, but leave a warning message

File: app/server/models/credit_transfer.py [hunk 9]
```diff
@@ -256,16 +270,22 @@ def resolve_as_complete(self, batch_uuid=None):
...
         if self.transfer_status not in [None, TransferStatusEnum.PENDING]:
-            raise Exception(f'Transfer resolve function called multiple times for transaciton {self.id}')
+            raise Exception(f'Resolve called multiple times for transfer {self.id}')
 
         if self.fiat_ramp and self.transfer_type in [TransferTypeEnum.DEPOSIT, TransferTypeEnum.WITHDRAWAL]:
             self.fiat_ramp.resolve_as_rejected()
 
         self.resolved_date = datetime.datetime.utcnow()
         self.transfer_status = TransferStatusEnum.REJECTED
+        self.update_balances()
```
Comment (enjeyw) [inline app/server/models/credit_transfer.py:280]: update balances after resolution, always

File: app/server/models/credit_transfer.py [hunk 10]
```diff
@@ -380,3 +401,14 @@ def __init__(self,
 
         self.append_organisation_if_required(self.recipient_transfer_account.organisation)
         self.append_organisation_if_required(self.sender_transfer_account.organisation)
+
+        if require_sufficient_balance and not self.check_sender_has_sufficient_balance():
```
Comment (enjeyw) [inline app/server/models/credit_transfer.py:405]: just check a credit transfer having sufficient balance by default, since it's kinda the norm

File: app/server/models/credit_transfer.py [hunk 11]
```diff
@@ -380,3 +401,14 @@ def __init__(self,
...
+            message = "Sender {} has insufficient balance. Has {}, needs {}.".format(
+                self.sender_transfer_account,
+                self.sender_transfer_account.balance,
+                self.transfer_amount
+            )
+            self.resolve_as_rejected(message)
+            raise InsufficientBalanceError(message)
+
+        self.update_balances()
```
Comment (enjeyw) [inline app/server/models/credit_transfer.py:414]: always update the balance cache after creating a transfer

File: app/server/models/exchange.py [hunk 12]
```diff
@@ -233,7 +228,9 @@ def exchange_from_amount(
             to_token,
             sender_transfer_account=find_transfer_accounts_with_matching_token(self.exchange_contract, to_token),
             recipient_user=user,
-            transfer_type=TransferTypeEnum.EXCHANGE, transfer_mode=transfer_mode
+            transfer_type=TransferTypeEnum.EXCHANGE,
+            transfer_mode=transfer_mode,
+            require_sufficient_balance=False
```
Comment (enjeyw) [inline app/server/models/exchange.py:233]: these are just sprinkled in whereever we explicitly weren't checking for sufficient balances previously

File: app/server/api/credit_transfer_api.py [hunk 13]
```diff
@@ -145,11 +145,21 @@ def put(self, credit_transfer_id):
...
+            if action == 'COMPLETE':
+                credit_transfer.resolve_as_complete_and_trigger_blockchain()
+
+            elif action == 'REJECT':
+                credit_transfer.resolve_as_rejected()
+
+        except Exception as e:
+
+            db.session.commit()
```
Comment (tristanhcole) [inline app/server/api/credit_transfer_api.py:157]: why db.session.commit() on a 400? To submit the error to the db?

File: app/server/api/credit_transfer_api.py [hunk 14]
```diff
@@ -487,6 +497,7 @@ def post(self):
                     transfer_type=TransferTypeEnum.PAYMENT,
                     sender_user=maybe_sender_user,
                     recipient_user=maybe_recipient_user,
+                    require_sufficient_balance=False
```
Comment (tristanhcole) [inline app/server/api/credit_transfer_api.py:500]: seems to be a weird, out of scope addition? @enjeyw please clarify

File: app/server/api/vendor_payout_api.py [hunk 15]
```diff
@@ -65,58 +67,74 @@ def post(self):
...
+                    send_user=v.primary_user,
+                    sender_transfer_account=v,
+                    transfer_mode=TransferModeEnum.INTERNAL,
+                    require_sender_approved=False,
+                    automatically_resolve_complete=False,
+                )
+
+                db.session.flush()
+
+                withdrawals.append(transfer)
```
Comment (tristanhcole) [inline app/server/api/vendor_payout_api.py:109]: would this be called again if using `relist_existing`?

File: app/server/api/vendor_payout_api.py [hunk 16]
```diff
@@ -136,62 +154,69 @@ def post(self):
...
-                t.sender_transfer_account.primary_user.last_name,
-                t.created,
-                t.transfer_type.value,
-                cents_to_dollars(t.transfer_amount),
-                t.transfer_status.value
+                tid,
+                t and t.sender_transfer_account.id,
+                t and t.sender_transfer_account.primary_user.phone,
+                t and t.sender_transfer_account.primary_user.first_name,
+                t and t.sender_transfer_account.primary_user.last_name,
```
Comment (tristanhcole) [inline app/server/api/vendor_payout_api.py:214]: if t isn't defined does a `None` value cause an error?

File: app/server/api/vendor_payout_api.py (hunk 7 above)
Comment (tristanhcole) [reply app/server/api/vendor_payout_api.py:189]: This is _wayyy more solid_

File: app/server/models/credit_transfer.py (hunk 9 above)
Comment (tristanhcole) [reply app/server/models/credit_transfer.py:280]: can you explain the reasoning as to why we would want to do this?

Comment (tristanhcole) [reply app/server/models/credit_transfer.py:280]: i.e. why weren't we doing this before?

File: app/server/api/credit_transfer_api.py (hunk 14 above)
Comment (tristanhcole) [reply app/server/api/credit_transfer_api.py:500]: ok makes sense now as it's been added through out the PR

File: app/server/api/vendor_payout_api.py [hunk 17]
```diff
@@ -65,58 +67,74 @@ def post(self):
...
+
+                db.session.flush()
+
+                withdrawals.append(transfer)
+
+            for w in withdrawals:
+                writer.writerow([
+                    v.id,
+                    v.primary_user.phone,
+                    v.primary_user.first_name + ' ' + v.primary_user.last_name,
```
Comment (tristanhcole) [inline app/server/api/vendor_payout_api.py:115]: I get this error when first or last name is null - maybe worth checking it exists or doing `str()` to ensure a 500 isn't accidently returned
```
	<title>TypeError: can only concatenate str (not &quot;NoneType&quot;) to str // Werkzeug Debugger</title>
```

File: app/test_app/functional/api/test_credit_transfer_api.py [hunk 18]
```diff
@@ -50,7 +50,7 @@ def test_create_credit_transfer(test_client, authed_sempo_admin_user, create_tra
     recipient_user_id = recipient_user_id_accessor(create_transfer_account_user)
 
     if transfer_type == 'PAYMENT' and sender_user_id:
-        create_transfer_account_user.transfer_account.set_balance_offset(10000)
+        create_transfer_account_user.transfer_account.set_balance_offset(1000000)
```
Comment (tristanhcole) [inline app/test_app/functional/api/test_credit_transfer_api.py:53]: ?????

**Review by tristanhcole — APPROVED** [review]
Tested and looks good to me! Nice work improving exception handling.

A few small questions that would be good to resolve ...

File: app/test_app/functional/api/test_credit_transfer_api.py (hunk 18 above)
Comment (tristanhcole) [reply app/test_app/functional/api/test_credit_transfer_api.py:53]: why is this off by two zero's now?

File: app/server/api/vendor_payout_api.py (hunk 16 above)
Comment (enjeyw) [reply app/server/api/vendor_payout_api.py:214]: No it just leaves a blank space

File: app/server/api/vendor_payout_api.py (hunk 17 above)
Comment (enjeyw) [reply app/server/api/vendor_payout_api.py:115]: ohhh great catch. will add

File: app/test_app/functional/api/test_credit_transfer_api.py (hunk 18 above)
Comment (enjeyw) [reply app/test_app/functional/api/test_credit_transfer_api.py:53]: oh it's just giving the test account more funds to play with - previous parts of the test were allowing the balance to dip into negative, where is the new code is more strict about allowing that sort of thing

File: app/server/models/credit_transfer.py (hunk 9 above)
Comment (enjeyw) [reply app/server/models/credit_transfer.py:280]: oh because previously pending transactions didn't change the balance, so there was no need to update after

File: app/server/api/credit_transfer_api.py (hunk 13 above)
Comment (enjeyw) [reply app/server/api/credit_transfer_api.py:157]: if a txn fails completion due to a limit, it's rejected and saved

File: app/server/api/vendor_payout_api.py [hunk 19]
```diff
@@ -26,8 +26,10 @@ def post(self):
         # Process post data
         post_data = request.get_json()
         account_ids = []
+        relist_existing = True
         if post_data:
             account_ids = post_data.get('accounts', [])
+            relist_existing = post_data.get('relist_existing', True)
 
         if not isinstance(account_ids, list):
```
Comment (tristanhcole) [inline app/server/api/vendor_payout_api.py:35]: IF a CONTRACT or ORGANISATION transfer account ID is passed back, then the `vendor.primary_user.has_vendor_role` on Line 44 throws a 500 error, as it is not caught by the SQL query on Line 42
```
File "/Users/tristancole/PycharmProjects/SempoBlockchain/app/server/api/vendor_payout_api.py", line 44, in post
    if not vendor.primary_user.has_vendor_role:
AttributeError: 'NoneType' object has no attribute 'has_vendor_role'
```
